pkg/services/ngalert/state/historian/loki.go

```diff
@@ -30,11 +36,113 @@ func (h *RemoteLokiBackend) TestConnection() error {
 	return h.client.ping()
 }
 
-func (h *RemoteLokiBackend) RecordStatesAsync(ctx context.Context, _ *models.AlertRule, _ []state.StateTransition) {
+func (h *RemoteLokiBackend) RecordStatesAsync(ctx context.Context, rule *models.AlertRule, states []state.StateTransition) {
 	logger := h.log.FromContext(ctx)
-	logger.Debug("Remote Loki state history backend was called with states")
+	streams := h.buildStreams(rule, states, logger)
+	h.recordStreamsAsync(ctx, streams, logger)
 }
 
 func (h *RemoteLokiBackend) QueryStates(ctx context.Context, query models.HistoryQuery) (*data.Frame, error) {
 	return data.NewFrame("states"), nil
 }
+
+func (h *RemoteLokiBackend) buildStreams(rule *models.AlertRule, states []state.StateTransition, logger log.Logger) []stream {
+	buckets := make(map[string][]row) // label repr -> entries
+	for _, state := range states {
+		if !shouldRecord(state) {
+			continue
+		}
+
+		labels := removePrivateLabels(state.State.Labels)
+		labels["orgID"] = fmt.Sprint(rule.OrgID)
+		labels["ruleUID"] = fmt.Sprint(rule.UID)
+		labels["group"] = fmt.Sprint(rule.RuleGroup)
+		labels["folderUID"] = fmt.Sprint(rule.NamespaceUID)
```

Comment:
nit: Do not have constants for these labels? They seem key enough to warrant their own.